vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.ts

```diff
@@ -270,18 +274,26 @@ import { GridFlowSelectionColumn } from "./vaadin-grid-flow-selection-column.js"
             if (sameLevelPage === null) {
               continue;
             }
-            firstNeededPage = Math.min(firstNeededPage, sameLevelPage);
-            lastNeededPage = Math.max(lastNeededPage, sameLevelPage);
+            pageRange[0] = Math.min(pageRange[0] ?? sameLevelPage, sameLevelPage);
+            pageRange[1] = Math.max(pageRange[1] ?? sameLevelPage, sameLevelPage);
+          }
+
+          // When the viewport doesn't contain the requested page or it doesn't contain any items from
+          // the requested level at all, it means that the scroll position has changed while fetchPage
+          // was debounced. For example, it can happen if the user scrolls the grid to the bottom and
+          // then immediately back to the top. In this case, the request for the last page will be left
+          // hanging. To avoid this, as a workaround, we reset the range to only include the requested page
+          // to make sure all hanging requests are resolved. After that, the grid requests the first page
+          // or whatever in the viewport again.
+          if (pageRange.some((p) => p === null) || page < pageRange[0] || page > pageRange[1]) {
+            pageRange = [page, page];
```

Comment:
**note:** Despite the workaround, there may be still situations when certain requests will end up hanging. For example, it can happen when the user scrolls from index range `[0, 10]` to index range `[10,  20]` and `size` suddenly changes to `10` before the debouncer timeout. In this case, the adjusted requested page will point to the end of the already loaded range `[0, 10]`, so the method will skip loading but the requests for `[10, 20]` still exist.

Reply:
To work this around, I considered doing a data communicator reset when the connector decides that the fetch range has not changed. However, when I tried that, I sometimes observed multiple round-trips, which made me afraid that it can result in a performance degradation given how convoluted the connector is at the moment.